File: library/utils/src/main/java/com/example/utils/format/NumFormat.java
package com.example.utils.format;

import android.text.TextUtils;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;

/**
 * 数字格式化工具类
 */
public class NumFormat {

	private volatile static NumFormat instance;

	// 保证单一实例
	private volatile static DecimalFormat decimalFormat;

	private NumFormat() {
		decimalFormat = new DecimalFormat();
	}

	public static NumFormat get() {
		if (instance == null) {
			synchronized (NumFormat.class) {
				if (instance == null) {
					instance = new NumFormat();
				}
			}
		}
		return instance;
	}

    /**
     * 四舍五入
     * @param pattern
     * @param value
     * @return
     */
    public String formatCeiling(String pattern, float value) {
        String valueStr = String.valueOf(value);
        if (TextUtils.isEmpty(pattern)) return valueStr;
        try {
            decimalFormat.applyPattern(pattern);
            decimalFormat.setRoundingMode(RoundingMode.CEILING);
            BigDecimal bd = new BigDecimal(String.valueOf(value));
            return decimalFormat.format(bd);
        } catch (IllegalArgumentException e) {
            return valueStr;
        }
    }

    /**
     * 非四舍五入
     * @param pattern
     * @param value
     * @return
     */
    public String formatFloor(String pattern, float value) {
        String valueStr = String.valueOf(value);
        if (TextUtils.isEmpty(pattern)) return valueStr;
        try {
            decimalFormat.applyPattern(pattern);
            decimalFormat.setRoundingMode(RoundingMode.FLOOR);
            BigDecimal bd = new BigDecimal(String.valueOf(value));
            return decimalFormat.format(bd);
        } catch (IllegalArgumentException e) {
            return valueStr;
        }
    }

}

File: library/utils/src/main/java/com/example/utils/format/NumFormat.kt
package com.example.utils.format

import android.text.TextUtils
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat

/**
 * 数字格式化工具类
 */
class NumFormat private constructor() {
    /**
     * 四舍五入
     * @param pattern
     * @param value
     * @return
     */
    fun formatCeiling(pattern: String?, value: Float): String {
        val valueStr = value.toString()
        return if (TextUtils.isEmpty(pattern)) valueStr else try {
            decimalFormat.applyPattern(pattern)
            decimalFormat.roundingMode = RoundingMode.CEILING
            val bd = BigDecimal(value.toString())
            decimalFormat.format(bd)
        } catch (e: IllegalArgumentException) {
            valueStr
        }
    }

    /**
     * 非四舍五入
     * @param pattern
     * @param value
     * @return
     */
    fun formatFloor(pattern: String?, value: Float): String {
        val valueStr = value.toString()
        return if (TextUtils.isEmpty(pattern)) valueStr else try {
            decimalFormat.applyPattern(pattern)
            decimalFormat.roundingMode = RoundingMode.FLOOR
            val bd = BigDecimal(value.toString())
            decimalFormat.format(bd)
        } catch (e: IllegalArgumentException) {
            valueStr
        }
    }

    companion object {
        @Volatile
        private var instance: NumFormat? = null

        // 保证单一实例
        @Volatile
        private lateinit var decimalFormat: DecimalFormat
        @JvmStatic
        fun get(): NumFormat? {
            if (instance == null) {
                synchronized(NumFormat::class.java) {
                    if (instance == null) {
                        instance = NumFormat()
                    }
                }
            }
            return instance
        }
    }

    init {
        decimalFormat = DecimalFormat()
    }
}